File: java/base/src/main/java/vsdk/toolkit/raycasting/photonMap/SampleGrid2D.java
/**
Class for doing multidimensional discrete sampling.
Grid values are doubles. Grid domain is [0,1]^dim
*/

package vsdk.toolkit.raycasting.photonMap;

import vsdk.toolkit.common.linealAlgebra.Numeric;

public class SampleGrid2D {
    private int xSections;
    private int ySections;
    private double[] values;
    private double[] ySums;  // Sum of y columns for faster sampling
    private double totalSum; // Normalisation factor

    private int valIndex(int i, int j) {
        return i * ySections + j;
    }

    public SampleGrid2D(int xSectionsParam, int ySectionsParam) {
        totalSum = 0.0;
        xSections = xSectionsParam;
        ySections = ySectionsParam;

        values = new double[xSections * ySections];
        ySums = new double[xSections];

        init();
    }

    public void init() {
        int index;

        index = 0;

        for ( int i = 0; i < xSections; i++ ) {
            ySums[i] = 0.0;
            for ( int j = 0; j < ySections; j++ ) {
                values[index++] = 0.0;
            }
        }

        totalSum = 0.0;
    }

    // Add a contribution to a certain grid element
    public void add(double x, double y, double value) {
        // Precondition: 0 <= x < 1 en 0 <= y < 1

        int xIndex;
        int yIndex;

        xIndex = (int)(x * xSections);
        yIndex = (int)(y * ySections);

        if ( xIndex == xSections ) {
            xIndex--;
        }  // x or y seem to be able to be 1
        if ( yIndex == ySections ) {
            yIndex--;
        }  // x or y seem to be able to be 1

        values[valIndex(xIndex, yIndex)] += value;
        ySums[xIndex] += value;
        totalSum += value;
    }

    // Ensure there are no zero value entries.
    // A small percentage of the totalPower is added to empty grid elements
    public void EnsureNonZeroEntries() {
        int index;
        // Add 3% of the average value to empty grid elements
        double fraction = 0.03 * totalSum / (xSections * ySections);
        double threshold = 1e-10 * totalSum;

        index = 0; // ! index is correlated with i,j in for loops

        for ( int i = 0; i < xSections; i++ ) {
            for ( int j = 0; j < ySections; j++ ) {
                if ( values[index] < threshold ) {
                    values[index] += fraction;
                    ySums[i] += fraction;
                    totalSum += fraction;
                }
                index++;
            }
        }
    }

    // Sample a point in a grid element. 'Random' values x and y get rescaled
    // so that they sample a uniform point in the selected grid element.
    // Note that their range is not [0,1] anymore, but smaller
    // probabilityDensityFunction for sampling this point is filled in.
    public void sample(double[] x, double[] y, double[] probabilityDensityFunction) {
        int xIndex;
        int yIndex;
        double[] xPdf = new double[1];
        double[] yPdf = new double[1];

        if ( totalSum < Numeric.EPSILON ) {
            // No significant data in table, use uniform sampling
            probabilityDensityFunction[0] = 1.0;
            return;
        }

        // Choose x row
        xIndex = DiscreteSampling.sample(ySums, totalSum, x, xPdf);

        // Choose y column
        double[] row = new double[ySections];
        System.arraycopy(values, xIndex * ySections, row, 0, ySections);
        yIndex = DiscreteSampling.sample(row, ySums[xIndex], y, yPdf);

        probabilityDensityFunction[0] = xPdf[0] * yPdf[0];

        // Rescale: x and y are in [0,1[ now we need to sample
        // grid element (xIndex, yIndex) uniformly

        double range;

        range = 1.0 / xSections;
        x[0] = (x[0] + xIndex) * range;
        probabilityDensityFunction[0] /= range;

        range = 1.0 / ySections;
        y[0] = (y[0] + yIndex) * range;
        probabilityDensityFunction[0] /= range;  // Uniform sampling: pdf = 1/A(xi,yi) * p(xi) * p(yi)
    }
}

File: typescript/base/src/main/vsdk/toolkit/raycasting/photonMap/SampleGrid2D.ts
/**
Class for doing multidimensional discrete sampling.
Grid values are doubles. Grid domain is [0,1]^dim
*/

import { Numeric } from "../../common/linealAlgebra/Numeric";
import { DiscreteSampling } from "./DiscreteSampling";

export class SampleGrid2D {
  private xSections: number;
  private ySections: number;
  private values: number[];
  private ySums: number[];
  private totalSum: number;

  private valIndex(i: number, j: number): number {
    return i * this.ySections + j;
  }

  public constructor(xSectionsParam: number, ySectionsParam: number) {
    this.totalSum = 0.0;
    this.xSections = xSectionsParam;
    this.ySections = ySectionsParam;

    this.values = new Array<number>(this.xSections * this.ySections).fill(0.0);
    this.ySums = new Array<number>(this.xSections).fill(0.0);

    this.init();
  }

  public init(): void {
    let index = 0;

    for (let i = 0; i < this.xSections; i++) {
      this.ySums[i] = 0.0;
      for (let j = 0; j < this.ySections; j++) {
        this.values[index++] = 0.0;
      }
    }

    this.totalSum = 0.0;
  }

  public add(x: number, y: number, value: number): void {
    let xIndex = globalThis.Math.trunc(x * this.xSections);
    let yIndex = globalThis.Math.trunc(y * this.ySections);

    if (xIndex === this.xSections) {
      xIndex--;
    }
    if (yIndex === this.ySections) {
      yIndex--;
    }

    this.values[this.valIndex(xIndex, yIndex)] += value;
    this.ySums[xIndex] += value;
    this.totalSum += value;
  }

  public EnsureNonZeroEntries(): void {
    let index = 0;
    const fraction = 0.03 * this.totalSum / (this.xSections * this.ySections);
    const threshold = 1e-10 * this.totalSum;

    for (let i = 0; i < this.xSections; i++) {
      for (let j = 0; j < this.ySections; j++) {
        if (this.values[index] < threshold) {
          this.values[index] += fraction;
          this.ySums[i] += fraction;
          this.totalSum += fraction;
        }
        index++;
      }
    }
  }

  public sample(x: number[], y: number[], probabilityDensityFunction: number[]): void {
    let xIndex: number;
    let yIndex: number;
    const xPdf = [0.0];
    const yPdf = [0.0];

    if (this.totalSum < Numeric.EPSILON) {
      probabilityDensityFunction[0] = 1.0;
      return;
    }

    xIndex = DiscreteSampling.sample(this.ySums, this.totalSum, x, xPdf);

    const row = new Array<number>(this.ySections).fill(0.0);
    for (let j = 0; j < this.ySections; j++) {
      row[j] = this.values[this.valIndex(xIndex, j)];
    }
    yIndex = DiscreteSampling.sample(row, this.ySums[xIndex], y, yPdf);

    probabilityDensityFunction[0] = xPdf[0] * yPdf[0];

    let range = 1.0 / this.xSections;
    x[0] = (x[0] + xIndex) * range;
    probabilityDensityFunction[0] /= range;

    range = 1.0 / this.ySections;
    y[0] = (y[0] + yIndex) * range;
    probabilityDensityFunction[0] /= range;
  }
}

